File: backend-csharp/Models/Subject.cs
namespace LmsApi.Models;

public class Subject
{
    public int Id { get; set; }
    public int CourseId { get; set; }
    public Course Course { get; set; } = null!;
    public string Name { get; set; } = string.Empty;
    public string Slug { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }

    public List<Question> Questions { get; set; } = new();
    public List<QuizSubject> QuizSubjects { get; set; } = new();
}

File: backend-node/src/models/Subject.ts
import mongoose, { Document, Schema, Types } from "mongoose";
import { slugify } from "../utils/Slugify";

export interface ISubject extends Document {
  course: Types.ObjectId;
  name: string;
  slug: string;
  description: string;
  createdAt: Date;
  updatedAt: Date;
}

const subjectSchema = new Schema<ISubject>(
  {
    course: { type: Schema.Types.ObjectId, ref: "Course", required: true },
    name: { type: String, required: true, unique: true },
    slug: { type: String, required: true, unique: true },
    description: { type: String, default: "" },
  },
  { timestamps: true }
);

function assignSlug(this: ISubject, next: () => void): void {
  if (this.isModified("name") || !this.slug) {
    this.slug = slugify(this.name);
  }
  next();
}

subjectSchema.pre("validate", assignSlug);

export default mongoose.model<ISubject>("Subject", subjectSchema);
